samza-core/src/main/java/org/apache/samza/zk/ZkBarrier.java

```diff
@@ -0,0 +1,214 @@
+ * the barrier, the creator marks the barrier as done which signals the end of barrier.
+ * The creator of the barrier can invoke {@link #expire(String)} to indicate that the barrier has timed_out or
+ * is no longer valid.
+ *
+ * The caller can listen to events associated with the barrier by registering a {@link ZkBarrierListener}.
+ *
+ * Zk Tree Reference:
+ * /barrierRoot/
+ *  |
+ *  |- barrier_{version1}/
+ *  |   |- barrier_state/
+ *  |   |  ([DONE|TIMED_OUT])
+ *  |   |- barrier_participants/
+ *  |   |   |- {id1}
+ *  |   |   |- {id2}
+ *  |   |   |-  ...
+ */
+public class ZkBarrier {
+  private final static Logger LOG = LoggerFactory.getLogger(ZkBarrier.class);
+  private static final String BARRIER_PARTICIPANTS = "/barrier_participants";
+  private static final String BARRIER_STATE = "/barrier_state";
+  private final ZkUtils zkUtils;
+  private final BarrierKeyBuilder keyBuilder;
+
+  public enum State {
+    TIMED_OUT, DONE
+  }
+
+  private ZkBarrierListener barrierListener;
+
+  public ZkBarrier(String barrierRoot, ZkUtils zkUtils, ZkBarrierListener barrierListener) {
+    if (zkUtils == null) {
+      throw new RuntimeException("Cannot operate ZkBarrierForVersionUpgrade without ZkUtils.");
+    }
+    this.zkUtils = zkUtils;
+    this.keyBuilder = new BarrierKeyBuilder(barrierRoot);
+    this.barrierListener = barrierListener;
+  }
+
+  /**
+   * Creates a shared barrier sub-tree in ZK
+   *
+   * @param version Version associated with the Barrier
+   * @param participants List of expected participated for this barrier to complete
+   */
+  public void create(final String version, List<String> participants) {
+    String barrierRoot = keyBuilder.getBarrierRoot();
+    zkUtils.makeSurePersistentPathsExists(new String[]{
+        barrierRoot,
+        keyBuilder.getBarrierPath(version),
+        keyBuilder.getBarrierParticipantsPath(version),
+        keyBuilder.getBarrierStatePath(version)});
+
+    // subscribe for participant's list changes
+    String barrierParticipantsPath = keyBuilder.getBarrierParticipantsPath(version);
+    LOG.info("Subscribing for child changes at " + barrierParticipantsPath);
+    zkUtils.getZkClient().subscribeChildChanges(barrierParticipantsPath, new ZkBarrierChangeHandler(version, participants));
+
+    if (barrierListener != null) {
+      barrierListener.onBarrierCreated(version);
+    }
+  }
+
+  /**
+   * Joins a shared barrier by registering under the barrier sub-tree in ZK
+   *
+   * @param version Version associated with the Barrier
+   * @param participantId Identifier of the participant
+   */
+  public void join(String version, String participantId) {
+    String barrierDonePath = keyBuilder.getBarrierStatePath(version);
+    zkUtils.getZkClient().subscribeDataChanges(barrierDonePath, new ZkBarrierReachedHandler(barrierDonePath, version));
+
+    // TODO: Handle ZkNodeExistsException
```

Comment:
What do we have to do in order to handle this exception? Not sure if it is complicated. If yes, is it desirable to defer it? If not, can we either do it right away or get rid of the _todo_?